ENH: umath: ensure ufuncs are well-defined with memory overlapping inputs

pv opened the pull request:
Make ufuncs and ufunc operations to make temporary copies if input and output operands overlap in memory. Currently (cf. gh-6272, gh-1683) output from such operations is undefined. With this PR, the result is the same as if the operands were non-overlapping.

Fixes gh-1683

Adds new flags to NpyIter, and one new API function is added to deal with MapIter.

This PR does not fix up overlap handling in array assignment, that needs to be done later (and is slightly more hairy --- some common overlapping cases are in fact already handled). Unlike adding overlap-if-copy flag to NpyIter, it appears a similar flag would not be as useful in MapIter, as there are many special-cased code paths, and MapIter is not always made aware of all arrays operated on.

- [x] reread whole patch
- [x] re-check test coverage, and whether some of the tests are unnecessary

pv committed:
ENH: NpyIter: add a flag to handle read/write operand overlap

Add a new NPY_ITER_COPY_IF_OVERLAP iterator flag to NpyIter, which
instructs it to check if read operands overlap with write operands in
memory, and make temporary copies to eliminate detected overlap.

Thanks to Sebastian Berg.

pv committed:
BUG: core: fix use of wrong output array in iterator_loop

These can differ if the iterator decides to make temporary copies of
output arrays.

pv committed:
ENH: core: add overlap detection logic to EQUIVALENTLY/TRIVIALLY_ITERABLE loops

These loops iterate over whole arrays in "trivial" order, so that it is
possible to reason about the data dependency.

pv committed:
ENH: umath: turn on overlap detection + copying in ufunc basic calls

After this commit, all code paths in PyUFunc_GeneralizedFunction and
PyUFunc_GenericFunction perform input/output operand memory overlap
detection, and make copies to eliminate it.

pv committed:
TST: core: add tests to check that basic ufunc calls deal with memory overlap

pv committed:
MAINT: core: add a missing compilation dependency to setup.py

pv committed:
STY: core: change test docstring to comment, to prettify nose output

pv committed:
BUG: core/ufunc: fix assumptions about iterator in execute_fancy_ufunc_loops

Because the wheremask loop does not write to all elements of the array, it
needs to consider output arrays as READWRITE, in case COPY_IF_OVERLAP makes
temporary copies.

As the iterator may make temporary copies of arrays, make sure to pass those to
__array_prepare__ instead of the originals.

pv committed:
TST: core: add tests for memory overlap in ufunc reductions

pv committed:
ENH: umath: turn on overlap detection + copying in ufunc reductions

pv committed:
ENH: core: handle memory overlap in ufunc.at

This adds a new method PyArray_MapIterArrayCopyIfOverlap to the API.

pv committed:
BUG: umath: in reduceat, arrays must be copied on overlap even if they are the same

pv committed:
STY: core: remove unused variable

charris commented on numpy/core/src/private/lowlevel_strided_loops.h:
```diff
@@ -673,6 +691,60 @@ npy_bswap8_unaligned(char * x)
+                                          PyArray_ITEMSIZE(arr)))
+
+static NPY_INLINE int
+PyArray_EQUIVALENTLY_ITERABLE_OVERLAP_OK(PyArrayObject *arr1, PyArrayObject *arr2,
```
Seems rather large for inline ;)

pv replied:
OTOH, I don't really want to put it in a .c file since the logic is tightly coupled to the macros here. Removing the inline makes it emit compiler warnings. I could write this as a macro, but that's messy.

charris replied:
C++ beckons ;)

charris commented on numpy/core/tests/test_mem_overlap.py:
```diff
@@ -4,9 +4,11 @@
 
 import numpy as np
-from numpy.testing import run_module_suite, assert_, assert_raises, assert_equal
+from numpy.testing import (run_module_suite, assert_, assert_raises, assert_equal, assert_array_equal,
```
Line too long ?

charris commented on numpy/core/code_generators/numpy_api.py:
```diff
@@ -344,6 +344,8 @@
     # End 1.9 API
     'PyArray_CheckAnyScalarExact':          (300, NonNull(1)),
     # End 1.10 API
+    # End 1.11 API
```
Should update this now that 1.12 is branched. The cversion and C-API number also need updating, see documentation in `doc/HOWTO_RELEASE.rst.txt`.

charris commented:
Just starting to look at this.

pv commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -1777,22 +1781,30 @@ execute_fancy_ufunc_loop(PyUFuncObject *ufunc,
             NpyIter_Deallocate(iter);
             return -1;
         }
+        Py_DECREF(op_tmp);
```
Now that I look at this again, I guess `op_tmp` should be placed back into `op` instead of being discarded.

charris commented on doc/source/reference/c-api.iterator.rst:
```diff
@@ -591,6 +619,15 @@ Construction and Destruction
+
+            In the memory overlap checks done when ``NPY_ITER_COPY_IF_OVERLAP``
+            is specified, consider this array as overlapping even if it is
+            exactly the same as another array.
```
?. What other array? I think the operation here could be best explained with the help of some diagrams showing the data access and writebacks. The operation can be performed, and is valid, even if there is no overlap, just that it isn't necessary if no overlap.

charris replied:
I think this might be help clarify when the indexing is many to one in either array.

charris commented on doc/source/reference/c-api.iterator.rst:
```diff
@@ -461,6 +461,34 @@ Construction and Destruction
+            operation is the same as if all operands were copied.
+            In cases where copies would need to be made, **the result of the
+            computation may be undefined without this flag!**
+
```
This is only for two operands? For inplace operations, both inputs are read, but only one is written too. Or does this apply for three operands, two inputs and one output. Or is it more general than that?

pv replied:
It applies to any number of operands.

charris commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -722,6 +723,38 @@ prepare_index(PyArrayObject *self, PyObject *index,
+ * @returns 1 if memory overlap found, 0 if not.
+ */
+NPY_NO_EXPORT int
+index_has_memory_overlap(PyArrayObject *self,
```
Hmm, so you are actually checking indexes rather than the resulting view?

pv replied:
It's used to checking the overlap between the array operated on, and the index array used in `ufunc.at`, for example
```
x = np.arange(100)
np.invert.at(x[::-1], x)  # <- oob memory access without x.copy() of the second argument
```

charris commented:
Just a first read through. This will need a release note eventually.

charris commented:
Probably should bring in @seberg on this also.

pv commented on numpy/core/tests/test_mem_overlap.py:
```diff
@@ -518,5 +564,363 @@ def __array__(self):
+    Test ufunc call memory overlap handling
+    """
+
+    def check_unary_fuzz(self, operation, get_out_axis_size, dtype=np.int16,
```
I would need to check again if these tests cover all places where the copyifoverlap behavior is in, and/or whether some of the tests are unnecessary.

pv replied:
ok, checked --- removing any of `NPY_COPY_IF_OVERLAP` et al. causes some test to fail

juliantaylor commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -3635,6 +3654,7 @@ PyUFunc_Reduceat(PyUFuncObject *ufunc, PyArrayObject *arr, PyArrayObject *ind,
         /* In case COPY or UPDATEIFCOPY occurred */
         op[0] = NpyIter_GetOperandArray(iter)[0];
         op[1] = NpyIter_GetOperandArray(iter)[1];
+        op[2] = NpyIter_GetOperandArray(iter)[2];
```
why is this line added? a bug in the old code or due to the other changes?

pv replied:
I think it's a noop. Not necessary to add.

pv replied:
On second thought, I think it's useful to have here to ensure correctness, if we later on change NPY_ITER_COPY_IF_OVERLAP to make copies of also the input arrays.

pv committed:
ENH: core: change NPY_ITER_OVERLAP_NOT_SAME to NPY_ITER_OVERLAP_ALLOW_SAME

It's clearer to explicitly specify which operands are accessed
elementwise, rather than to list exceptions.

pv committed:
MAINT: bump API version

pv committed:
MAINT: drop NPY_INLINE

pv committed:
BUG: umath: correct treatment of __array_prepare__

prepare_ufunc_output is allowed to replace the array, because we check
that it's still a view. Also if COPY_IF_OVERLAP made updateifcopy of
op[i], the data will be reflected in the array from __array_prepare__.

The is-a-view check is only bare-bones, but it was so since 61d45eee

pv committed:
STY: break long line

pv committed:
DOC: update release notes vs. overlap detection

pv committed:
DOC: fix typo in comment

pv committed:
MAINT: core: fix compiler warnings

pv committed:
BUG: core: fix bug with zero strides in PyArray_EQUIVALENTLY_ITERABLE_OVERLAP_OK

Arrays with zero strides are never "ahead", as the same element is
reused multiple times.

The previous logic was incorrect eg. for stride1==0, stride2<0 and
arr1->data==arr2->data.

pv committed:
TST: core: streamline and extend test_mem_overlap.TestUFunc tests

Use itertools to go over all permutations rather than trying to do it
manually.

pv commented:
Fixed one bug an re-checked test coverage.

Here's a question: what should the following operation produce:
```
>>> import numpy as np
>>> from numpy.lib.stride_tricks import as_strided
>>> x = as_strided(np.array([1]), (2,), (0,))
>>> np.add(x, x.copy(), x)
# array([3, 3])   OR
# array([2, 2]), as in np.add(x.copy(), x.copy(), x)
```
There are some tests in `test_nditer.py` that expect that `y[...] += x` gives the [3,3] type result when `y` has zero strides. Can we chalk this off as relying on undefined behavior?

pv committed:
ENH: core: rename NPY_ITER_OVERLAP_ALLOW_SAME to NPY_ITER_OVERLAP_ASSUME_ELEMENTWISE

Make the flag name indicate more clearly on what is the precise
assumption made.

seberg commented:
It is a problem, short of special casing these some kind of overlaps as being OK, there is nothing we can do. But, I do expect I even have some very old dirty scripts that might break by this, which doesn't mean much, just saying that I do think that some breakage will probably occur in some weirder scripts.

However, the only possible way to move forward I see is to first of all hope that nothing really breaks, and if it does (or do we have to assume it will?), it would be starting off with a FutureWarning and doing the actual change later. I guess in principle the futurewarning could be given only in cases that were not complete bogus before/easy to abuse, which probably is limited to identical strides.

pv committed:
BUG: core/umath: fix cornercase for overlapping operands with internal overlap

Ensure that a copy is made when ufunc output operand overlaps with ufunc
input, and both have internal overlap. The heuristic applied when
iteration order is elementwise is OK only in the absence of internal
overlap.

pv commented:
I think the earlier checks showed that mainstream projects such as sklearn, pandas, scipy etc. do not really have overlapping arrays in their testsuites. So I think we may need to make the decision on what is a FutureWarning, without knowing example use in the wild.

What should be FutureWarning --- everything, or just ovelapping-output-has-internal-overlap case?
If we want to be conservative, we could just make this just emit futurewarnings first, and then turn it on for reals only in the next release. Or maybe also make it conditional on some compile-time flag such as the relaxed strides stuff was.

pv committed:
BUG: core/umath: handle __array_prepare__ correctly in ufunc iterator_loop

Ensure that results get written to the array returned by
__array_prepare__, if any. This necessitates calling
prepare_ufunc_output on the output arrays that the iterator may replace
with UPDATEIFCOPY'd ones.

Add a test to the test suite --- this code path was untested previously.

seberg commented:
Its cool this is getting close. A compile time flag may be an option. Overall, I don't think any major project is likely to run into this, and any project with a test suit should notice immidiately. I am a bit scared of breaking stuff, but on the other hand, I don't really believe we will break a lot (or if, the break is probably visible at least). To be honest, I think we probably had this discussion before, and the examples I thought of last time were incorrect ;).

Frankly, maybe I am optimistic, but I think we may just want to give it a shot....

mhvk commented:
I'm coming very late to this, but I'm worried about this test case:
```
       x = np.add(a, a, a)
       assert_(not np.shares_memory(x, a))  # returned array is always new
```
Previous it was guaranteed that if you passed in an `out` argument, that would be the one returned (indeed, there is currently #8416 to remind us that this is not the case for `MaskedArray`). Is this change really wanted? I thought the point was to make temporary copies, which are then stored back in the output array if needed. Does this change imply that `a += a` also returns a new array?

This example seems particularly strange, since it is one where overlap really isn't a problem and thus is more likely to be seen in the wild...

pv commented:
@mhvk: I don't think there are changes vs `__array_prepare__` behavior. (I tested vs astropy and all tests pass.) The comment "returned array is always new" is due to the specific `__array_prepare__` method always returning a new output array no matter what.

pv commented:
I'm actually not sure if `__array_prepare__` returning a fully new output array is fully indented behavior, as it's apparently not allowed with `where=`, see [current code in master branch](https://github.com/numpy/numpy/blob/e0cf8393d3b7e8bb6ae9f7c9871bf0a90e3b4896/numpy/core/src/umath/ufunc_object.c#L1799). It also appears gufuncs don't support array_prepare at all? All this is not really related to this PR, but this stuff looks a bit half-cooked (and I guess it will be obsoleted by numpy_ufunc_override, so no need to fix it...).

pv committed:
DOC: core/umath: make a comment in the test be less misleading

mhvk commented:
OK, I indeed also checked astropy with your PR, and things pass, plus the test case I had in mind. Sorry for not realising the test was more unusual in calling `np.array`... I actually don't think one is supposed to be able to return anything but a view in `__array_prepare__`...

mhvk commented:
From the documentation [1], it seems indeed that `__array_prepare__` should not touch/copy the output data (though another page [2] is a bit less explicit):
"In addition to __array_wrap__, which is called on the way out of the ufunc, there is also an __array_prepare__ method which is called on the way into the ufunc, after the output arrays are created but before any computation has been performed. The default implementation does nothing but pass through the array. __array_prepare__ should not attempt to access the array data or resize the array, it is intended for setting the output array type, updating attributes and metadata, and performing any checks based on the input that may be desired before computation begins. Like __array_wrap__, __array_prepare__ must return an ndarray or subclass thereof or raise an error."

[1] https://docs.scipy.org/doc/numpy/user/basics.subclassing.html
[2] https://docs.scipy.org/doc/numpy/reference/arrays.classes.html

njsmith commented:
Historically, we've said that stuff like `np.add(a, b, a)` where the overlap is confined to within each "core" operation is well-defined, and that everything else involving overlap between inputs and outputs, or between outputs and outputs (like the 0-strided output case), is undefined. Plus we've regularly broken the overlap cases over the last few years (e.g. due to implementing iteration order optimizations), which has caused a few people to complain but wasn't a disaster.

Doc issue: the current PR doesn't make clear what happens to the cases that were always well-defined (`np.add(a, b, a)` etc.). I assume that they continue to be well-defined, and that they continue *not* to make a copy? This should be explicit though - if you take the current changes.rst text literally then I think it implies that they now make an (unnecessary) copy. I'm also not 100% clear on what this patch does for weirdnesses like `np.add(a, b, out=np.as_strided(np.array([1]), shape=(2,), strides=(0,))` -- I assume this still invokes undefined behavior? (I guess the ideal behavior here would just be an error, b/c what is that even supposed to mean.) It'd be nice if there were a bit of text in the docs that lays out all the different cases and how they're handled.

For the example
```python
>>> x = as_strided(np.array([1]), shape=(2,), strides=(0,))
>>> np.add(x, x.copy(), x)
```
then `[3, 3]` seems reasonable to me, if we even want to support it at all. (This is only well-defined because it happens that all the outputs are identical, right? `np.add(x, [1, 2], x)` is ill-defined regardless of this patch, because even `np.add([1, 1], [1, 2], x)` is already ill-defined.) I'm sorta feeling like we don't need any `FutureWarning`s here, because the "breakage" is no more than regularly happens due to like, memory layout changes. OTOH if we wanted to be conservative and issue `FutureWarning`s then that would be OK with me too, and I'd vote that instead of trying to be super-clever about detecting which cases to issue the warnings in, we just run the new logic for deciding when to copy and have a release or two that issues a `FutureWarning` in *every* case where the make-a-copy-logic activates.

njsmith commented:
(Also, this is really awesome and exciting!)

pv committed:
DOC: releases: clarify simple data dependency case in ufunc overlap discussion

pv commented:
The change here is that all ufuncs operations produce the same results as similar operation on arrays that have the same content but no read/write memory overlap. For overlapping arrays, copies are skipped if the data dependency is simple enough --- currently, this is only done in the "exactly same array" and "trivial 1d iteration with simple data dependence" cases. So yes, `np.add(a, b, a)` will not make a copy (unless `a` happens to have internal overlap in which case a copy is made).

I think the general case of zero-stride output that overlaps with inputs is undefined. The `[3,3]` indeed appears sort of reasonable because the arrays are the same. However, this would be an ugly corner case if it is left in IMHO (so it's currently removed), as it breaks the "`x += y` is the same as `x[...] = x + y`" rule.

Although we *could* also detect and handle the case where the only overlap is internal overlap in the write operand, I didn't try to change the behavior here, it is still undefined.

njsmith commented:
Oh, sorry, I misread on the `[3, 3]` thing - I meant that I don't see any problem with changing it, but got mixed up and thought `[3, 3]` was the "new" output. I should have said `[2, 2]` :-)

njsmith commented:
Okay, so to rephrase for my own understanding:

* It used to be that ufunc operations were ill-defined when there was overlap between one of the input arrays and one of the output arrays, *except* in the case where the overlap was only between scalars at identical indices.

* Now, ufunc operations are well-defined regardless of overlap between inputs and outputs. An operation `foo(a, b, out=(b, c))` is defined to be equivalent to the same operation where all the inputs are replaced by copies of them. However, in some cases we are able to optimize out these copies (this includes both the cases that were previously well-defined, but also some new cases like your example `np.add(x[1:], x[:-1], x[:-1])` that used to be ill-defined!). One can imagine further optimizations like this in the future, but since the semantics are now well-defined this would be purely an internal thing.

* Everything above only discussed overlap between an input array and an output array. There are two more cases to mention for completeness. For overlap between an input array and an input array (possibly itself), there's never been any problem and there still isn't -- `add(a, a.T)` or `add(zero_strided_a, b)` have always been well-defined and still are. For overlap between an output array and an output array (like `out=(a, a)`, or `out=zero_strided_a`, or ...), then it was always ill-defined, and remains ill-defined.

That all makes sense and sounds good to me! But...

> (unless `a` happens to have internal overlap in which case a copy is made)

...this line confuses me then :-). According to my analysis above, internal overlap shouldn't affect anything -- internal overlap in an input array doesn't matter, and internal overlap in an output array makes the whole operation ill-defined. So I must be missing something. When does internal overlap trigger a copy and why?

pv commented:
@njsmith: it's internal overlap in an output array that also overlaps with an input. I guess handling this case wouldn't be strictly necessary if we declare it an undefined operation. It just happens that numpy test suite had two cases of this (trying to be clever and writing reductions using `+=` --- but this being unrelated to what was being tested).

njsmith commented:
It seems to me that those tests are just wrong though then... I mean, we can make the copy, it doesn't matter much either way I guess, but either way the final result is still going to be dependent on the internal ufunc iteration order, right?

pv commented:
It's order-independent (but probably not buffering independent), as it's
a simple scalar reduction case.

njsmith commented:
What I mean is, if we have `annoying = np.as_strided(np.array([1]), shape=(2,), strides=(0,)`, then `np.add([1, 1], [1, 2], out=annoying)` could reasonably give either `[2, 2]` or `[3, 3]`, depending on whether it writes the output array from front-to-back or back-to-front. We're writing both 2 and 3 to the same location, but we don't guarantee which will win.

pv commented:
That's correct, but I think the specific reduction case in the tests I
wrote about was order-independent as it's `np.add([1, 2], annoying,
out=annoying)`. In any case, those tests are now rewritten, so if the
understanding is that this is in general undefined behavior, there are
no further changes required.

njsmith commented:
Oh, right, now I understand! ...this rings a bell, too. @seberg, weren't you telling me at some point about how there's some code inside numpy that relies on implementing reducations via zero-strided calls to the non-reduce loops? Would that be affected by this? (My guess is that whatever code is using this trick is "below" the overlap checking in the stack so unaffected by it, and I'm totally happy with this not being something we guarantee to end-users. Also hopefully no-one is writing horrible `as_strided` expressions as a "shorthand" instead of calling `np.sum`!)

pv commented:
The copies are triggered only if the iterator is given 
NPY_ITER_COPY_IF_OVERLAP flag, which I was careful to set only if I 
understand the code --- so it is not set for the reduction iterators, 
and the overlap issue is handled before the iterator.

pv committed:
DOC: fix wrong flag name docstring

pv committed:
BUG: multiarray: fix refcount error in error handling in PyArray_MapIterArrayCopyIfOverlap

charris commented:
@pv Do you feel that this is about ready?

pv committed:
BUG: core: fix refcount errors on error handling

pv commented:
@charris: I've read this several times through now, and only finding forgotten decref on error handling. I'm fairly confident now that it's what it says on the tin, and doesn't do anything more.

It's a bit ugly that this adds `PyArray_MapIterArrayCopyIfOverlap` to the API --- it's essentially a private function, as was `PyArray_MapIterArray` --- but the better way is not obvious. On the other hand, adding the new iterator flags appears to be a good way to do this, as it makes reasoning about the change easier, and the changes in ufuncs themselves are mostly minimal.

pv commented:
One additional thing is that this probably won't work on PyPy due to the use of updateifcopy, but I guess that's part of a bigger pypy issue that needs separate fixes (probably sprinkling `PyArray_FlushUpdateIfCopy` around) in more places than just this.

charris commented:
@mattip @rlamy Comments?

rlamy commented:
Well, I don't really understand the UPDATEIFCOPY issue, but since @mattip is on holiday, I'll comment anyway.

All the new tests appear to pass on both pypy and pypy3. I'm guessing that's because the copy cannot be seen by the RPython GC. So I think this is OK for us.

pv commented:
Yes, the updateifcopy arrays here are never passed back to Python. If 
cpyext is effectively refcounted in this case, then there should be no 
problem. I guess things would break also in other places if this wasn't 
the case, as the iterator uses updateifcopy temporaries also for other 
purposes.

charris commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -704,6 +705,38 @@ prepare_index(PyArrayObject *self, PyObject *index,
+    if (index_type & (HAS_FANCY | HAS_BOOL)) {
+        for (i = 0; i < num; ++i) {
+            if (indices[i].object != NULL && PyArray_Check(indices[i].object) &&
+                solve_may_share_memory(self, (PyArrayObject *)indices[i].object,
```
Should have extra indent here to separate the condition from the block.

charris replied:
Bit tricky with the long line, but it doesn't read well as is.

charris commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -704,6 +705,38 @@ prepare_index(PyArrayObject *self, PyObject *index,
+    }
+
+    if (extra_op != NULL && PyArray_Check(extra_op) &&
+        solve_may_share_memory(self, (PyArrayObject *)extra_op, 1) != 0) {
```
See above.

charris commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -3197,6 +3259,21 @@ PyArray_MapIterArray(PyArrayObject * a, PyObject * index)
 
+/*NUMPY_API
+ *
+ * Use advanced indexing to iterate an array. Please note
```
This needs to be documented somewhere public. It seems dangerous to expose an API that may change...

charris replied:
How could someone detect the change to this particular function?

pv replied:
This the is old comment for `PyArray_MapIterArray`, not added by me.

pv replied:
Anyway, removing it, since it's unlikely we will change this...

charris commented on numpy/core/src/multiarray/nditer_constr.c:
```diff
@@ -2708,6 +2710,80 @@ npyiter_allocate_arrays(NpyIter *iter,
     }
 
+    if (flags & NPY_ITER_COPY_IF_OVERLAP) {
+        /* Perform operand memory overlap checks if requested */
```
Could add more detail on what is being checked here, the notes below are a bit too local.

charris commented on numpy/core/src/multiarray/nditer_constr.c:
```diff
@@ -2708,6 +2710,80 @@ npyiter_allocate_arrays(NpyIter *iter,
+                 * make sense to go further.
+                 */
+                may_share_memory = solve_may_share_memory(
+                    op[iop], op[iother], 1);
```
Maybe extra indent here, or even break the args and align.

charris commented on numpy/core/src/private/lowlevel_strided_loops.h:
```diff
@@ -673,6 +691,63 @@ npy_bswap8_unaligned(char * x)
+        return 1;
+    }
+
+    /* Arrays overlapping in memory may be equivalently iterable
```
Multiline comment not correctly formatted ;) Also below.

charris commented on numpy/core/src/private/lowlevel_strided_loops.h:
```diff
@@ -673,6 +691,63 @@ npy_bswap8_unaligned(char * x)
+    }
+    else if (stride1 < 0) {
+        arr1_ahead = (stride1 <= stride2 &&
+                      (npy_uintp)PyArray_BYTES(arr1) <= (npy_uintp)PyArray_BYTES(arr2));
```
Hmm... Out of curiousity, did the compiler complain without the cast?

pv replied:
The casts are not needed (probably leftovers from earlier versions).

charris commented on numpy/core/src/umath/ufunc_object.c:
```diff
@@ -1730,12 +1764,17 @@ execute_fancy_ufunc_loop(PyUFuncObject *ufunc,
     for (i = nin; i < nop; ++i) {
+        /*
+         * Because we don't write to all elements, the output arrays
+         * must be considered READWRITE by the iterator.
```
Curious.

pv replied:
Updateifcopy copies all elements back, not only those for which wheremask is true, so the old content of the output array must be copied there.

charris commented:
Not done reviewing yet. Determining when a copy isn't needed for overlapping arrays seems tricky except in single operand the linear case. ISTR that the iterator may coalesce axes.

pv committed:
STY: core: fix up code style and add clearer comments

pv committed:
DOC: core: remove old comment saying public API may change

It's unlikely we will break backward compatibility, so better
not claim we will do it in the code.

charris commented:
I'm going to put this in so it will start getting more testing. I don't claim to have understood all the details, but I doubt I will discover any errors by inspection. The write up inth release notes is very nice, you should consider copying that over to the official documentation. There are still some style nits that I may attempt to clean up later, I would gently urge you to pay more attention to spaces around operators ('+') and line length, even though the latter may not always be as readable. 

Thanks for this work, Pauli.

seberg commented:
Wow, cool. Assuming nothing goes haywire, this is the next huge addition :)!
